File: src/UE4_Assets/FSkinWeightInfo.java
/**
 * 
 */
package UE4_Assets;

import UE4.FArchive;

/**
 * @author FunGames
 *
 */
public class FSkinWeightInfo {
	
	private byte[] boneIndex; //Size 8
	private byte[] boneWeight; //Size 8


	public byte[] getBoneIndex() {
		return boneIndex;
	}

	public byte[] getBoneWeight() {
		return boneWeight;
	}

	public FSkinWeightInfo(FArchive Ar, int maxInfluences) throws ReadException {
		if(maxInfluences > 8) {
			throw new ReadException("Max influences too high", Ar.Tell());
		}
		boneIndex = new byte[maxInfluences];
		for(int i=0;i<maxInfluences; i++) {
			boneIndex[i] = Ar.readUInt8();
		}
		boneWeight = new byte[maxInfluences];
		for(int i=0;i<maxInfluences; i++) {
			boneWeight[i] = Ar.readUInt8();
		}
	}
}

File: src/main/kotlin/me/fungames/jfortniteparse/ue4/assets/objects/meshes/FSkinWeightInfo.kt
package me.fungames.jfortniteparse.ue4.assets.objects.meshes

import me.fungames.jfortniteparse.ue4.reader.FArchive

class FSkinWeightInfo {
    companion object {
        private const val NUM_INFLUENCES_UE4 = 4
        private const val MAX_TOTAL_INFLUENCES_UE4 = 8
    }

    val boneIndex = UByteArray(NUM_INFLUENCES_UE4)
    val boneWeight = UByteArray(NUM_INFLUENCES_UE4)

    constructor()

    constructor(Ar: FArchive, numSkelCondition: Boolean) {
        val numSkelInfluences = if (numSkelCondition) MAX_TOTAL_INFLUENCES_UE4 else NUM_INFLUENCES_UE4
        if (numSkelInfluences <= boneIndex.size) {
            repeat(numSkelInfluences) { boneIndex[it] = Ar.readUInt8() }
            repeat(numSkelInfluences) { boneWeight[it] = Ar.readUInt8() }
        } else {
            val boneIndex2 = UByteArray(MAX_TOTAL_INFLUENCES_UE4)
            val boneWeight2 = UByteArray(MAX_TOTAL_INFLUENCES_UE4)
            repeat(numSkelInfluences) { boneIndex2[it] = Ar.readUInt8() }
            repeat(numSkelInfluences) { boneWeight2[it] = Ar.readUInt8() }

            // copy influences to vertex
            repeat(NUM_INFLUENCES_UE4) {
                boneIndex[it] = boneIndex2[it]
                boneWeight[it] = boneWeight2[it]
            }
        }
    }
}